src/scheduler.cc

```diff
@@ -14,35 +14,45 @@ namespace detail {
 	void abstract_scheduler::shutdown() { notify(scheduler_event_type::SHUTDOWN, 0); }
 
 	bool abstract_scheduler::is_idle() const noexcept {
-		std::lock_guard<std::mutex> lock(events_mutex);
-		return events.empty();
+		std::lock_guard lock(events_mutex);
+		return available_events.empty() && in_flight_count == 0;
 	}
 
 	void abstract_scheduler::schedule() {
-		std::unique_lock<std::mutex> lk(events_mutex);
-
 		while(true) {
-			// TODO: We currently operate in lockstep with the main thread. This is less than ideal.
-			events_cv.wait(lk, [this] { return !events.empty(); });
-			const auto event = events.front();
-			events.pop();
-
-			const task_id tid = event.data;
-			if(event.type == scheduler_event_type::TASK_AVAILABLE) {
-				naive_split_transformer naive_split(num_nodes, num_nodes);
-				ggen.build_task(tid, {&naive_split});
-				gsrlzr.flush(tid);
-			} else if(event.type == scheduler_event_type::SHUTDOWN) {
-				assert(events.empty());
-				return;
+			{
+				std::unique_lock lk(events_mutex);
+				events_cv.wait(lk, [this] { return !available_events.empty(); });
+				std::swap(available_events, in_flight_events);
+				in_flight_count = in_flight_events.size();
+			}
+
+			while(!in_flight_events.empty()) {
+				const auto event = std::move(in_flight_events.front());
+				in_flight_events.pop();
+
+				if(event.type == scheduler_event_type::SHUTDOWN) {
+					assert(in_flight_events.empty());
+					in_flight_count -= 1;
+					return;
+				}
+
+				if(event.type == scheduler_event_type::TASK_AVAILABLE) {
```

Comment:
This could be an assertion instead of an if, which would make it more obvious that no event is skipped – and by extension `in_flight_count` never misses an update.

Reply:
Right, changed.